kafka-client/src/test/java/dev/responsive/kafka/integration/OriginEventIntegrationTest.java

```diff
@@ -0,0 +1,413 @@
+
+  @RegisterExtension
+  static ResponsiveExtension EXTENSION = new ResponsiveExtension(StorageBackend.MONGO_DB);
+
+  private static final String INPUT_TOPIC = "input";
+  private static final String OUTPUT_TOPIC = "output";
+
+  private final Map<String, Object> responsiveProps = new HashMap<>();
+
+  private String name;
+  private TestLicenseServer licenseServer;
+
+  @BeforeEach
+  public void before(
+      final TestInfo info,
+      final Admin admin,
+      @ResponsiveConfigParam final Map<String, Object> responsiveProps,
+      final TestLicenseServer licenseServer
+  ) throws ExecutionException, InterruptedException {
+    name = IntegrationTestUtils.getCassandraValidName(info);
+    this.licenseServer = licenseServer;
+    this.responsiveProps.putAll(responsiveProps);
+
+    // report every commit, so we don't potentially miss origin events
+    this.responsiveProps.put(ResponsiveConfig.ORIGIN_EVENT_REPORT_INTERVAL_MS_CONFIG, 0);
+
+    final var result = admin.createTopics(
+        List.of(
+            new NewTopic(inputTopic(), Optional.of(2), Optional.empty()),
+            new NewTopic(inputTopicTable(), Optional.of(2), Optional.empty()),
+            new NewTopic(outputTopic(), Optional.of(1), Optional.empty())
+        )
+    );
+    result.all().get();
+  }
+
+  private String inputTopic() {
+    return name + "." + INPUT_TOPIC;
+  }
+
+  private String inputTopicTable() {
+    return name + ".table." + INPUT_TOPIC;
+  }
+
+  private String outputTopic() {
+    return name + "." + OUTPUT_TOPIC;
+  }
+
+  @ParameterizedTest
+  @ValueSource(booleans = {true, false})
+  public void shouldCountOriginEventsSimpleTopology(final boolean eos) throws Exception {
+    // Given:
+    responsiveProps.put(
+        StreamsConfig.PROCESSING_GUARANTEE_CONFIG,
+        eos ? StreamsConfig.EXACTLY_ONCE_V2 : StreamsConfig.AT_LEAST_ONCE
+    );
+
+    final int numEvents = 100;
+    final CountDownLatch latch = new CountDownLatch(numEvents);
+    final StreamsBuilder builder = new StreamsBuilder();
+    builder
+        .stream(inputTopic())
+        .peek((k, v) -> latch.countDown())
+        .to(outputTopic());
+
+    final List<KeyValueTimestamp<String, String>> inputs = IntStream.range(0, numEvents)
+        .mapToObj(i -> new KeyValueTimestamp<>("key" + i, String.valueOf(i), i))
+        .collect(Collectors.toList());
+
+    // When:
+    final var props = getDefaultMutablePropertiesWithStringSerdes(responsiveProps, name);
+    final KafkaProducer<String, String> producer = new KafkaProducer<>(props);
+    try (final var kafkaStreams = new ResponsiveKafkaStreams(builder.build(), props)) {
+      startAppAndAwaitRunning(Duration.ofSeconds(15), kafkaStreams);
+      pipeTimestampedRecords(producer, inputTopic(), inputs);
+      assertThat(latch.await(30, TimeUnit.SECONDS), Matchers.is(true));
+    }
+
+    // Then:
+    final String appId = (String) props.get(StreamsConfig.APPLICATION_ID_CONFIG);
+    IntegrationTestUtils.awaitCondition(
```

Comment:
why do we need to do this in `awaitCondition`? The app is shut down so whatever counts that the license server has should be reflected at this point right? That way we fail immediately rather than on a timeout if we misreport.